Update index.md (#23)
security-summer-school/essentials: data-representation/index.md

text2 = "SecuritySummmerScho"
text3 = "SecuritySummmerSchoo"
text4 = "SecuritySummmerSchool"

b64_text1 = (base64.b64encode(text1.encode('ascii'))).decode('ascii')
b64_text2 = (base64.b64encode(text2.encode('ascii'))).decode('ascii')
b64_text3 = (base64.b64encode(text3.encode('ascii'))).decode('ascii')
b64_text4 = (base64.b64encode(text4.encode('ascii'))).decode('ascii')

print(f"Plain text\t\tPT length\tBase64 text\t\t\tB64 length")
print(f"{text1}\t{len(text1)}\t\t{b64_text1}\t{len(b64_text1)}")
print(f"{text2}\t{len(text2)}\t\t{b64_text2}\t{len(b64_text2)}")
print(f"{text3}\t{len(text3)}\t\t{b64_text3}\t{len(b64_text3)}")
print(f"{text4}\t{len(text4)}\t\t{b64_text4}\t{len(b64_text4)}")
```

We get:
```html
Plain text              PT length       Base64 text                     B64 length
SecuritySummmerSch      18              U2VjdXJpdHlTdW1tbWVyU2No        24
SecuritySummmerScho     19              U2VjdXJpdHlTdW1tbWVyU2Nobw==    28
SecuritySummmerSchoo    20              U2VjdXJpdHlTdW1tbWVyU2Nob28=    28
SecuritySummmerSchool   21              U2VjdXJpdHlTdW1tbWVyU2Nob29s    28
```

Try decoding yourself!
```
SGVsbG8gZnJvbSB0aGUgRWFydGgtNjQgIQ==
```

> #### `Hashing`

Reviewer: Should this section contain more about security?
I would add something about Rainbow tables [1] ( awesome documentation), but I don't know if it's worth to be added here or in the Further reading section.

CC: @teodutu ( please give your input here)

[1]: https://kestas.kuliukas.com/RainbowTables/

> #### `Hashing`

Reviewer: ```suggestion
#### Hashing
```

Reviewer: Any reason to be quoted? Does it looks better on website?

d992ee83f3675b4021e0bb614021ffeb
```

2. Getting the hash (`sha-1` in this case) of a text using `Python`:

We will use the module `hashlib` in order to get hashes in Python:

```py
import hashlib

text = "some kind of text"
h = hashlib.sha1(text.encode('ascii'))

print(h.hexdigest())
```

This code would provide the following `sha-1` hash:
```
1145d8b71c84d1b0a717d36bd0cd6638697bd606
```

We can store passwords as MD5, SHA-1 or SHA-2 (most popular ones), so that in case of a security breach, the real passwords are still hidden.  
An unsecure and naive way of storing passwords is:

```
|  User   |      Password      |
|:-------:|:------------------:|
| Robert  | S3cur3.p@ssw0rd?   |
| Maria   | Palmademallorca123 |
| Ciprian | password           |
```

A much better approach is to store the passwords hashed, as mentioned above:
```
|  User   |             Password             |
|:-------:|:--------------------------------:|
| Robert  | 0daeb943b4cc762da2c541c7338c90b6 |
| Maria   | 9daa2547b9a0155841d7f653257e3762 |
| Ciprian | 5f4dcc3b5aa765d61d8327deb882cf99 |
```

## Summary

Data can be represented in many ways, with different purposes.

Some useful information from today:

> - Hashes cannot be reverted

Reviewer: I would change the text from here because there are some hashes that can be reverted and are still used with different use cases. (e.g. store image on database using md5)

- Base64 is encoding, not encrypting
- Computers love binary
 
Some useful commands from today:
- Convert Binary to Decimal - `int(n,2)`
- Find the ASCII code - ord('X')
- Find a character based on its ASCII code - chr(88)

> ## Activities

Reviewer: Don't forget to update this section.

- Basic knowledge of Python (as seen in the [first session](../welcome-to-linux/))
- Numbers

## What is Data?

Data is information.
This plain text is data, but more than that, data can be encoded and represented in many ways.  
Generally, we represent data in a suitable format for our specific purpose.  
For example, if we want the most basic way to encode data, the one that computers "think" in, we'll use `Binary`.  
Sometimes, we need our data to not be confused with something else, so, for example we encode `Binary` to `Base-64` in order to get the information on the other side of the wire uncorrupted.

## Data Formats

### Numeral Systems
Numeral Systems are a method of representing numbers by mathematical combinations of symbols.

![Numeral Systems 1](./assets/numeral_systems1.svg)

Humans prefer the `decimal` numeral system (also know as Base-10), since it provides better readability.  
Therefore, written software (i.e. code) is mostly `Base-10`, as humans write software far more than they build hardware.

Computers, on the other hand, use binary (or Base-2), the numerical system that uses two digits (0 and 1), which are also known as `bits` and `bytes` (1 byte = 8 bits)

> ![Binary Meme](./assets/binary_meme.svg)

Reviewer 2: :heart:

12  0C	FF          38	26  &           64  40  @           90  5A  Z           116 74	t
13  0D	CR          39	27  '           65  41  A           91  5B  [           117 75	u
14  0E	SO          40	28  (           66  42  B           92  5C  \           118 76	v
15  0F	SI          41	29  )           67  43  C           93  5D  ]           119 77	w
16  10	DLE         42	2A  *           68  44  D           94  5E  ^           120 78	x
17  11	DC1         43	2B  +           69  45  E           95  5F  _           121 79	y
18  12	DC2         44	2C  ,           70  46  F           96  60  `           122 7A	z
19  13	DC3         45	2D  -           71  47  G           97  61  a           123 7B	{
20  14	DC4         46	2E  .           72  48  H           98  62  b           124 7C	|
21  15	NAK         47	2F  /           73  49  I           99  63  c           125 7D	}
22  16	SYN         48	30  0           74  4A  J           100 64  d           126 7E	~
23  17	ETB         49	31  1           75  4B  K           101 65  e           127 7F
24  18	CAN         50	32  2           76  4C  L           102 66  f 
25  19	EM          51	33  3           77  4D  M           103 67  g 
```

> In terms of storage efficiency, we can encode
> 
> `UTF-8` for ASCII text (English and other Western languages)
> 
> `UTF-16` for non-ASCII text (Chinese and other Asian languages)

Reviewer 2: ```suggestion
In terms of storage efficiency, we can encode:
- `UTF-8` for ASCII text (English and other Western languages)
- `UTF-16` for non-ASCII text (Chinese and other Asian languages)
```

## Introduction

In today's session we'll discuss about data and the many ways we can represent it.

## Reminders and Prerequisites

For this session you'll need:

> - Basic knowledge of Python (as seen in the [first session](https://security-summer-school.github.io/essentials/welcome-to-linux/))

Reviewer 2: ```suggestion
- Basic knowledge of Python (as seen in the [first session](..//welcome-to-linux/))
```

Use relative links. If we change our domain name to `sss-upb.ro` for example, relative links would still work.

- Numbers

> - 

Reviewer 2: ```suggestion
```

## What is Data?

Data is information.

> This plain text is data, but more than that, data can be encoded and represented in many ways.
> 
> Generally, we represent data in a suitable format for our specific purpose.
> 
> For example, if we want the most basic way to encode data, the one that computers "think" in, we'll use `Binary`.
> 

Reviewer 2: ```suggestion
This plain text is data, but more than that, data can be encoded and represented in many ways.
Generally, we represent data in a suitable format for our specific purpose.
For example, if we want the most basic way to encode data, the one that computers "think" in, we'll use `Binary`.
```

Sometimes, we need our data to not be confused with something else, so, for example we encode `Binary` to `Base-64` in order to get the information on the other side of the wire uncorrupted.

## Data Formats

### Number Systems
Number Systems are a method of representing numbers by mathematical combinations of symbols.
![Number Systems](./assets/number_systems.svg)

#### `Decimal`

> Decimal or Base-10 uses 10 digits (0-9). It's the `human` way to represent numbers.

Reviewer 2: ```suggestion
Decimal or Base-10 uses 10 digits (0-9). It's the **human** way to represent numbers.
```

#### `Binary `
Binary or Base-2 only uses two digits (0 and 1).

0 and 1 are called `bits` and `8 bits = 1 byte`.

Computers store data in bites and bytes.

But why do computers do this?

Hardware prefers them, since it is easier to associate with electrical signals:

![Data Signals](./assets/data_signals.svg)

#### `Hexadecimal`
Hexadecimal or Base-16 uses 10 digits (0-9) and 6 alphabet letters (A-F). We can look at hexadecimal data as a shortcut for the binary system.

Let's say we have `0b10101001` (`10101001`).

I assume you can safely say that since we have 8 bits, it will be `< 256`.

Its hexadecimal form is `0xa9` (`a9`).

Thus, if we want to convert it to `decimal`, instead of doing 8 steps:

$(1 × 2^7) + (0 × 2^6) + (1 × 2^5) + (0 × 2^4) + (1 × 2^3) + (0 × 2^2) + (0 × 2^1) + (1 × 2^0) = 169$

we only do 2 steps:

$(a × 16^1) + (9 × 16^0) = 169$

> TODO(Cod Python)

Reviewer 2: I would arrange this differently. Start with a drawing similar to the sketch below:
```
=== software === <- humans
----------------
=== hardware ===
```
First, say that humans prefer base 10 for easy readability. Therefore, software also uses base 10 (mostly), as humans write software far more than they build hardware.

Then say that hardware likes the binary representation, which you've already done. Then introduce hexadecimal as a midpoint between the 2 systems: both readable and tightly correlated to the binary representation.

#### `Octal`
Octal or Base-8 uses 8 digits (0-7). It is the least popular of the aforementioned 4, but an interesting use of it is in the Unix File Permissions system:

![Unix File Permissions](./assets/unix_file_permissions.svg)

> 

Reviewer 2: Expand this section a bit and say what these permissions mean, especially for folders, where the "execute" permission allows traversal.

12  0C	FF          38	26  &           64  40  @           90  5A  Z           116 74	t
13  0D	CR          39	27  '           65  41  A           91  5B  [           117 75	u
14  0E	SO          40	28  (           66  42  B           92  5C  \           118 76	v
15  0F	SI          41	29  )           67  43  C           93  5D  ]           119 77	w
16  10	DLE         42	2A  *           68  44  D           94  5E  ^           120 78	x
17  11	DC1         43	2B  +           69  45  E           95  5F  _           121 79	y
18  12	DC2         44	2C  ,           70  46  F           96  60  `           122 7A	z
19  13	DC3         45	2D  -           71  47  G           97  61  a           123 7B	{
20  14	DC4         46	2E  .           72  48  H           98  62  b           124 7C	|
21  15	NAK         47	2F  /           73  49  I           99  63  c           125 7D	}
22  16	SYN         48	30  0           74  4A  J           100 64  d           126 7E	~
23  17	ETB         49	31  1           75  4B  K           101 65  e           127 7F
24  18	CAN         50	32  2           76  4C  L           102 66  f 
25  19	EM          51	33  3           77  4D  M           103 67  g 

> ```

Reviewer 2: The whole table is nice, but follow it up by some detailed information such as the ASCII codes for letters and digits, what converting between lowercase and uppercase means and some Python demos/exercises with ASCII encodings.

One challenge idea can be to write a flag such as `SSS{whatever}` but with the string inside `{...}` encoded in ASCII and the task would be to decode the concatenated ASCII codes to get the string.

In terms of storage efficiency, we can encode

`UTF-8` for ASCII text (English and other Western languages)

> `UTF-16` for non-ASCII text (Chinese and other Asian languages)

Reviewer 2: Do go into some more details regarding those. Or at least put some links about `utf-8` and `utf-16` in the "Further Reading" section.

#### `Base64`

Base64 is a way of representing binary data in sequences of 24 bits that can be represented by 4 Base64 digits.

SGVsbG8gZnJvbSB0aGUgRWFydGgtNjQgIQ==

> 

Reviewer 2: - Add Python demo. This topic is heavily misunderstood.
- Explain the algorithm
- Say that there is also base32

## Data Manipulation

We can manually change the way data is represented, so that it will be easier to read or structure.

For large chunks of data, this could take a long time, so, in order for us to be efficient, we will have to use certain automated ways of manipulating the information.

Tools that can help us achieve this goal are ranging from programming and scripting languages, to programs like MS Excel.

> For this session's purpose, we will mainly focus on Python as already seen in [Session 1](https://security-summer-school.github.io/essentials/welcome-to-linux/)

Reviewer 2: ```suggestion
For this session's purpose, we will mainly focus on Python as already seen in [Session 1](../welcome-to-linux/)
```

## Reminders and Prerequisites

For this session you'll need:
- Basic knowledge of Python (as seen in the [first session](..//welcome-to-linux/))
- Numbers

## What is Data?

Data is information.
This plain text is data, but more than that, data can be encoded and represented in many ways.
Generally, we represent data in a suitable format for our specific purpose.
For example, if we want the most basic way to encode data, the one that computers "think" in, we'll use `Binary`.
Sometimes, we need our data to not be confused with something else, so, for example we encode `Binary` to `Base-64` in order to get the information on the other side of the wire uncorrupted.

## Data Formats

### Numeral Systems
Numeral Systems are a method of representing numbers by mathematical combinations of symbols.

![Numeral Systems 1](./assets/numeral_systems1.svg)

> Humans prefer the `decimal` numeral system (also know as Base-10), since it provides better readability, hence the software of computers is mostly `Base-10`, as humans write software far more than they build hardware.

Reviewer 2: ```suggestion
Humans prefer the `decimal` numeral system (also know as Base-10), since it provides better readability.
Therefore, written software (i.e. code) is mostly `Base-10`, as humans write software far more than they build hardware.
```

> Computers, on the other hand, use binary (or Base-2), the numerical system that uses two digits (0 and 1), which are also known as `bits` and `bytes` (1 byte = 8 bits)

Reviewer 2: Don't mix bits and bytes or you'll create confusion. Talk about bytes when you describe encodings (ASCII, UTF, etc).

## Further Reading

### Data Manipulation
We can manually change the way data is represented, so that it will be easier to read or structure.

For large chunks of data, this could take a long time, so, in order for us to be efficient, we will have to use certain automated ways of manipulating the information.

Tools that can help us achieve this goal are ranging from programming and scripting languages, to programs like MS Excel.  
A useful Python tool for this purpose is `pandas`:

[pandas (Python tool for Data Manipulation)(1)](https://pandas.pydata.org/docs/getting_started/)  
[pandas (Python tool for Data Manipulation)(2)](https://www.w3schools.com/python/pandas/default.asp)

### Hashes and their vulnerabilities
[CrackStation](https://crackstation.net/)  
[Hash Collision Wikipedia](https://en.wikipedia.org/wiki/Hash_collision)  

> [Shattered](https://shattered.io/)

Reviewer: I would also add hashtoolkit here.

https://hashtoolkit.com/
